source/browseMode.py

```diff
@@ -2001,6 +2040,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
+	) -> "textInfos.TextInfo.TextWithFieldsT":
+		"""
+		This function calls TextInfo.getTextWithFields(), and then processes fiields in the following way:
+		1. Highlighted(marked) text is currently reported as Role.MARKED_CONTENT, and not formatChange.
+		For ease of further handling we create a new boolean format field "marked"
+		and set its value according to presence of Role.MARKED_CONTENT.
+		2. Then we drop all control fields, leaving only formatChange fields and text.
+		"""
+		stack = [{}]
+		result = []
+		for field in info.getTextWithFields():
+			if isinstance(field, textInfos.FieldCommand):
+				if field.command == 'controlStart':
+					style = {**stack[-1]}
+					if field.field['role'] == controlTypes.Role.MARKED_CONTENT:
+						style['marked'] = True
+					stack.append(style)
+				elif field.command == 'controlEnd':
+					del stack[-1]
+				elif field.command == 'formatChange':
+					field.field = {
+						k: v
+						for k, v in {**field.field, **stack[-1]}.items()
+						if k in self.STYLE_ATTRIBUTES
+					}
+					result.append(field)
+				else:
+					raise RuntimeError()
+			elif isinstance(field, str):
+				result.append(field)
+			else:
+				raise RuntimeError
+		return result
+
+	def _iterTextStyle(
+			self,
+			kind: str,
+			direction: str = "next",
+			pos: textInfos.TextInfo = None
+	) -> Generator[TextInfoQuickNavItem, None, None]:
+		sameStyle = kind == 'sameStyle'
+		initialTextInfo = pos.copy()
+		initialTextInfo.collapse()
+		result = initialTextInfo.move(textInfos.UNIT_CHARACTER, 1, endPoint='end')
+		if result == 0:
+			result = initialTextInfo.move(textInfos.UNIT_CHARACTER, -1, endPoint='start')
+			if result == 0:
+				# Translators: Error message for same/different style QuickNav command
+				ui.message(_("Cannot determine current style"))
+				raise RuntimeError
+		styles = self._extractStyles(initialTextInfo)
+		if (
+			len(styles) == 0
+			or not isinstance(styles[0], textInfos.FieldCommand)
+			or styles[0].command != 'formatChange'
+		):
+			# Translators: Error message for same/different style QuickNav command
+			ui.message(_("Cannot determine current style"))
+			raise RuntimeError
+		initialStyle = styles[0]
+
+		firstParagraph = True
+		paragraph = pos.copy()
+		tmpInfo = pos.copy()
+		tmpInfo.expand(textInfos.UNIT_PARAGRAPH)
+		paragraph.setEndPoint(tmpInfo, which='endToEnd' if direction == 'next' else 'startToStart')
+		while True:
+			if not paragraph.isCollapsed:
+				styles = self._extractStyles(paragraph)
+				firstStyleWithinParagraph = True
+				iterationRange = range(len(styles)) if direction == 'next' else range(len(styles) - 1, -1, -1)
+				for i in iterationRange:
+					if not isinstance(styles[i], textInfos.FieldCommand):
+						continue
+					if (
+						(styles[i].field == initialStyle.field) == sameStyle
+						and (
+							not firstStyleWithinParagraph
+							or not firstParagraph
+						)
+					):
+						# Found text that matches desired style!
+						startOffset = sum([

```

Comment:
You are assuming here that the number of characters in a textInfo is equal to the number of characters in a Python string, right?
This is not true, as many Offset based textInfo implementations including virtual buffers work with UTF16 characters. There, emoji are treated as two characters, whereas they are one character in Python.
You should be able to convert between widestring and Python offsets with the tools in the textUtils module.